src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseMessage.cs

```diff
@@ -0,0 +1,184 @@
+
+        private CoseHeaderMap _protectedHeaders;
+        private CoseHeaderMap _unprotectedHeaders;
+        public CoseHeaderMap ProtectedHeader => _protectedHeaders;
+        public CoseHeaderMap UnprotectedHeader => _unprotectedHeaders;
+
+        internal CoseMessage(CoseHeaderMap protectedHeader, CoseHeaderMap unprotectedHeader, byte[]? content, byte[] signature, byte[] encodedProtectedHeader)
+        {
+            _content = content;
+            _signature = signature;
+            _protectedHeaderAsBstr = encodedProtectedHeader;
+            _protectedHeaders = protectedHeader;
+            _unprotectedHeaders = unprotectedHeader;
+        }
+
+        // Sign and MAC also refer to the content as payload.
+        // Encrypt also refers to the content as cyphertext.
+        public ReadOnlyMemory<byte>? Content
+        {
+            get
+            {
+                if (_content  != null)
+                {
+                    return _content;
+                }
+
+                return null;
+            }
+        }
+
+        public static CoseSign1Message DecodeSign1(byte[] cborPayload)
+        {
+            try
+            {
+                var reader = new CborReader(cborPayload);
+                CborTag? tag = DecodeTag(reader);
+                if (tag != null && tag != Sign1Tag)
+                {
+                    throw new CryptographicException(SR.Format(SR.DecodeSign1IncorrectTag, tag));
+                }
+
+                CoseSign1Message message = DecodeCoseSign1Core(reader);
+                return reader.BytesRemaining == 0 ? message : throw new CryptographicException(SR.Format(SR.DecodeSign1ErrorWhileDecoding, SR.DecodeSign1MesageContainedTrailingData));
+            }
+            catch (Exception ex) when (ex is CborContentException or InvalidOperationException)
+            {
+                throw new CryptographicException(SR.DecodeSign1ErrorWhileDecodingSeeInnerEx, ex);
+            }
+        }
+
+        private static CborTag? DecodeTag(CborReader reader)
+        {
+            return reader.PeekState() switch
+            {
+                CborReaderState.Tag => reader.ReadTag(),
+                _ => null
+            };
+        }
+
+        private static CoseSign1Message DecodeCoseSign1Core(CborReader reader)
+        {
+            int? arrayLength = reader.ReadStartArray();
+            if (arrayLength != 4)
+            {
+                throw new CryptographicException(SR.Format(SR.DecodeSign1ErrorWhileDecoding, SR.DecodeSign1ArrayLengthMustBeFour));
+            }
+
+            var protectedHeader = new CoseHeaderMap();
+            DecodeProtectedBucket(reader, protectedHeader, out byte[] protectedHeaderAsBstr);
+            protectedHeader.IsReadOnly = true;
+
+            var unprotectedHeader = new CoseHeaderMap();
+            DecodeUnprotectedBucket(reader, unprotectedHeader);
+
+            CoseSign1Message.ThrowIfDuplicateLabels(protectedHeader, unprotectedHeader);
```

Comment:
Is `ThrowIfDuplicateLabels` a Sign1 behavior, or a CoseMessage behavior?  Feels base-classish to me.

Reply:
Right, it is CoseMessage, will pass the helper to the base class.